src/main/scala/dsls/regex/RegularExpression.scala

```diff
@@ -8,11 +9,29 @@ package dsls.regex
  */
 
 /** The top of a class hierarchy that encodes regular expressions. */
-abstract class RegularExpression {
+abstract class RegularExpression{
   /** returns true if the given string matches this regular expression */
   def matches(string: String) = RegexMatcher.matches(string, this)
-}
 
+  // union operator 
+  def ||(in: RegularExpression) = Union(in, this)
+
+  // concatenation operator
+  def ~(in: RegularExpression) = Concat(in, this)
+
+  // star operator (zero or more repititions)
+  def <*> = Star(this)
+
+  // plus operator (one or more repititions)
+  def <+> = Concat(this, Star(this))
+
+  // the repitition operator (n repititions)
+  def apply(n: Int): RegularExpression = {
```

Comment:
I implemented all of these in either the exact or very similar ways. I was impressed by how simple it looks after its written even though I thought it was weird to write.